src/build_vehicle_gui.cpp

```diff
@@ -269,20 +269,25 @@ static bool EnginePowerVsRunningCostSorter(const EngineID &a, const EngineID &b)
 {
 	const Engine *e_a = Engine::Get(a);
 	const Engine *e_b = Engine::Get(b);
+	uint p_a = e_a->GetPower();
+	uint p_b = e_b->GetPower();
+	Money r_a = e_a->GetRunningCost();
+	Money r_b = e_b->GetRunningCost();
+	/* Check if running cost is zero in one or both engines */
+	if (r_a == 0) {
+		if (r_b == 0) {
+			if (p_a == p_b) return EngineNumberSorter(a,b);
+			return _engine_sort_direction ? p_a >= p_b : p_a < p_b;
+		}
+		return !_engine_sort_direction;
+	}
+	if (r_b == 0 ) return _engine_sort_direction;
 
-	/* Here we are using a few tricks to get the right sort.
-	 * We want power/running cost, but since we usually got higher running cost than power and we store the result in an int,
-	 * we will actually calculate cunning cost/power (to make it more than 1).
-	 * Because of this, the return value have to be reversed as well and we return b - a instead of a - b.
-	 * Another thing is that both power and running costs should be doubled for multiheaded engines.
-	 * Since it would be multiplying with 2 in both numerator and denominator, it will even themselves out and we skip checking for multiheaded. */
-	Money va = (e_a->GetRunningCost()) / max(1U, (uint)e_a->GetPower());
-	Money vb = (e_b->GetRunningCost()) / max(1U, (uint)e_b->GetPower());
-	int r = ClampToI32(vb - va);
-
+	double va = (double)p_a / (double)r_a;
+	double vb = (double)p_b / (double)r_b;
```

Comment:
Could probably use a comment justifying the use of `double` here

Reply:
... i meant the part where this is GUI code and isn't restricted to integer-maths

Reply:
... that's still not what i meant.

the requirement for integer-maths is not about performance at all, but about reproduceablity. float-maths can have different results depending on processor architecture and other implementation details (rounding errors), so the game logic cannot use it for anything that must be synchronized in multiplayer.

Reply:
oh, right. yeah but because it is a GUI/client only thing it shouldn't have an effect in syncing, right?

Reply:
exactly, that is what the comment should say.

Reply:
I am sorry. Nobody said this specifically so i thought that was they meant by that.